### PR TITLE
fix: avoid additional array allocation in host to device transfer

### PR DESCRIPTION
**Pull request by Alexandr-Solovev**
## Description

PR introduces fix for memcpy function in oneDAL. It saves memory and improve the performance of transferring data from host to device and back and gives an opportunity to use memcpy directly.

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

**Comment by Alexandr-Solovev** [comment]
Please dont merge, until the additional check on local machines

**Comment by Alexandr-Solovev** [comment]
> Looks like #2375 was just missing the wait_and_throw() :(

not sure, looks like in all places where the function is called there is additional wait_and_throw()

File: cpp/oneapi/dal/table/backend/convert.cpp
Comment (Vika-F) [inline cpp/oneapi/dal/table/backend/convert.cpp:241]: I guess there should be something like this. Because `int32` and `float` will have `src_element_size_in_bytes` == `dst_element_size_in_bytes`.

```suggestion
    if (src_type == dst_type &&
```

File: cpp/oneapi/dal/table/backend/convert.cpp
Comment (Vika-F) [inline cpp/oneapi/dal/table/backend/convert.cpp:307]: Same comment here.

File: cpp/oneapi/dal/table/backend/homogen_kernels.cpp
Comment (icfaust) [inline cpp/oneapi/dal/table/backend/homogen_kernels.cpp:134]: Just because I am not familiar, it looks like just removing the if and relying on the for loop (a simplification here).

File: cpp/oneapi/dal/table/backend/convert.cpp
Comment (icfaust) [inline cpp/oneapi/dal/table/backend/convert.cpp:306]: If allocation exists, skip the convert_convert vector and uses memcpy_host2usm

File: cpp/oneapi/dal/table/backend/convert.cpp
Comment (icfaust) [inline cpp/oneapi/dal/table/backend/convert.cpp:230]: collect more information about the destination to see if a shortcut can be used.

**Review by icfaust — APPROVED** [review]
Made comments help to me understand. Would it be worth just adding some? Otherwise good to go by me (if my comments in the review are correct)

File: cpp/oneapi/dal/table/backend/convert.cpp
Comment (Alexandr-Solovev) [reply cpp/oneapi/dal/table/backend/convert.cpp:230]: Exactly!

**Review by samir-nasibli — COMMENTED** [review]
Indeed good suggestion! Thank you @Alexandr-Solovev 
Could you please share perf improvements?

**Comment by ethanglaser** [comment]
Are you still seeing seg faults when experimenting locally or has this issue been resolved?

**Pull request merged**
